restapi/src/main/java/io/stargate/web/docsapi/service/JsonConverterService.java

```diff
@@ -0,0 +1,290 @@
+
+        if (!shouldWrite) {
+          markFullPathAsDead(parentPath, p, deadLeaves);
+          break;
+        }
+
+        if (endOfPath) {
+          boolean shouldBeArray = isArray && !ref.isArray() && !writeAllPathsAsObjects;
+          if (i == 0 && shouldBeArray) {
+            doc = mapper.createArrayNode();
+            ref = doc;
+            pathWriteTimes.put(parentPath, rowWriteTime);
+          } else if (i != 0 && shouldBeArray) {
+            markObjectAtPathAsDead(ref, parentPath, deadLeaves);
+            ref = changeCurrentNodeToArray(row, parentRef, i);
+            pathWriteTimes.put(parentPath, rowWriteTime);
+          } else if (i != 0 && !isArray && !ref.isObject()) {
+            markArrayAtPathAsDead(ref, parentPath, deadLeaves);
+            ref = changeCurrentNodeToObject(row, parentRef, i, writeAllPathsAsObjects);
+            pathWriteTimes.put(parentPath, rowWriteTime);
+          }
+          leaf = p;
+          break;
+        }
+
+        JsonNode childRef;
+
+        if (isArray && !writeAllPathsAsObjects) {
+          if (!ref.isArray()) {
+            if (i == 0) {
+              doc = mapper.createArrayNode();
+              ref = doc;
+              pathWriteTimes.put(parentPath, rowWriteTime);
+            } else {
+              markObjectAtPathAsDead(ref, parentPath, deadLeaves);
+              ref = changeCurrentNodeToArray(row, parentRef, i);
+              pathWriteTimes.put(parentPath, rowWriteTime);
+            }
+          }
+
+          int index = Integer.parseInt(p.substring(1, p.length() - 1));
+
+          ArrayNode arrayRef = (ArrayNode) ref;
+
+          int currentSize = arrayRef.size();
+          for (int k = currentSize; k < index; k++) arrayRef.addNull();
+
+          if (currentSize <= index) {
+            childRef = nextIsArray ? mapper.createArrayNode() : mapper.createObjectNode();
+            arrayRef.add(childRef);
+          } else {
+            childRef = arrayRef.get(index);
+
+            // Replace null from above (out of order)
+            if (childRef.isNull()) {
+              childRef = nextIsArray ? mapper.createArrayNode() : mapper.createObjectNode();
+            }
+
+            arrayRef.set(index, childRef);
+          }
+          parentRef = ref;
+          ref = childRef;
+        } else {
+          childRef = ref.get(p);
+          if (childRef == null) {
+            childRef =
+                nextIsArray && !writeAllPathsAsObjects
+                    ? mapper.createArrayNode()
+                    : mapper.createObjectNode();
+
+            if (!ref.isObject()) {
+              markArrayAtPathAsDead(ref, parentPath, deadLeaves);
+              ref = changeCurrentNodeToObject(row, parentRef, i, writeAllPathsAsObjects);
+              pathWriteTimes.put(parentPath, rowWriteTime);
+            }
+
+            ((ObjectNode) ref).set(p, childRef);
+          }
+          parentRef = ref;
+          ref = childRef;
```

Comment:
These statements are the same on the if branch at lines 130 and 131. They could be extracted from this inner loop.